app/services/doc_auth/error_generator.rb

```diff
@@ -189,14 +200,42 @@ def get_error_messages(liveness_enabled, response_info)
         end
       end
 
-      portrait_match_results = response_info[:portrait_match_results] || {}
-      if liveness_enabled && portrait_match_results.dig(:FaceMatchResult) != 'Pass'
-        errors[SELFIE] << Errors::SELFIE_FAILURE
+      selfie_error = get_selfie_error(liveness_enabled, response_info)
+      if liveness_enabled && selfie_error
+        errors[SELFIE] << selfie_error
       end
 
       errors
     end
 
+    def get_selfie_error(liveness_enabled, response_info)
+      # The part of the response that contains information about the selfie
+      portrait_match_results = response_info[:portrait_match_results] || {}
+      # The overall result of the selfie, 'Pass' or 'Fail'
+      face_match_result = portrait_match_results.dig(:FaceMatchResult)
+      # The reason for failure (if it failed), also sometimes contains success info
+      face_match_error = portrait_match_results.dig(:FaceErrorMessage)
+
+      # No error if liveness is not enabled or if there's no failure
+      if !liveness_enabled || face_match_result == 'Pass'
+        return nil
+      end
+
+      # Error when the image on the id does not match the selfie image, but the image was acceptable
+      if face_match_result == 'Fail' &&
+         face_match_error == "FaceErrorMessage: 'Successful. Liveness: Live'"
+        return Errors::SELFIE_FAILURE
+      end
+      # Error when the image on the id is poor quality
+      if face_match_result == 'Fail' && face_match_error == 'Liveness: PoorQuality'
```

Comment:
I am wondering whether we need to check face_match_result here, since we already get an error on the liveness check of the selfie, itself is enough to indicate error.

Reply:
I think this is also addressed [here](https://github.com/18F/identity-idp/pull/9941/commits/ee025ced8cd8932d0f75b426b3c9e8890c652dc3).